src/advanced/service-testing.md

```diff
@@ -1,155 +1,260 @@
+            "/v1/second_endpoint",
+        )
+    }
+
+    fn post_private_data(&self, data: &PrivateData)
+        -> PostPrivateDataResponse
+    {
+        self.post(
+            ApiKind::Service("my_service"),
+            "v1/third_endpoint",
+            data,
+        )
+    }
+}
 
-    // Check if the node has sent the "Connect" message
-    s.send(s.a(2), Connect::new(&s.p(0), s.a(0), s.time(), s.s(0)));
+#[test]
+fn my_api_test() {
+    let mut testkit = TestKitBuilder::validator()
+        .with_service(MyService::new())
+        .create();
+    fill_storage_with_data(&mut testkit);
+    // Check API responses
+    let api = testkit.api();
+    assert_eq!(
+        api.get_public_data(),
+        ApiResponsePublicData::new(...),
+    );
+    ...
 }
+
+// Other tests...
 ```
 
-## Service Testing
+## Advanced Usage
 
-To test a certain set of the services, one should pass the services list to the
-sandbox constructor:
+The testkit allows to test more complex behaviors of Exonum services,
+such as getting data from external sources and reconfiguring the service.
 
-```rust
-let s = sandbox_with_services(vec![Box::new(TimestampingService::new()),
-                                   Box::new(ConfigUpdateService::new())]);
-```
+### Oracles Testing
 
-A service should be tested by committing the service transaction and observing
-subsequent changes in the blockchain and the storage state.
+The *oracle* is a service which can produce transactions
+with external data after the commit of the block.
+[The Bitcoin anchoring service](bitcoin-anchoring.md) is an example of an oracle.
+Just like a real Exonum node, the testkit maintains a pool of unconfirmed transactions
+(aka the *mempool*). Thus, transactions created by the oracle service
+during the [`handle_commit`][Service.handle_commit] execution
+will be stored in `TestKit` memory pool, and can be verified accordingly.
 
-!!! note
-    Transaction constructors are service-specific.
+```rust
+// Create testkit with the service which creates transaction
+// with the height of the latest committed block after commit.
+let mut testkit = TestKitBuilder::validator()
+    .with_service(HandleCommitService)
+    .create();
+
+// Call the `handle_commit` event.
+testkit.create_block();
+
+// Check that `handle_commit` has been invoked
+// at the correct height.
+let tx = TxAfterCommit::new_with_signature(
+    Height(1),
+    &Signature::zero(),
+);
+assert!(testkit.mempool().contains_key(&tx.hash()));
+```
 
 !!! tip
-    See source code for more details on how sandbox is used for testing
-    [the configuration update service](https://github.com/exonum/exonum-configuration/blob/master/sandbox_tests/src/lib.rs)
-    and [the anchoring service](https://github.com/exonum/exonum-btc-anchoring/tree/master/sandbox_tests/tests).
+    In order to invoke a `handle_commit` event, you need to create a block
+    with one of `create_block*` methods of the testkit.
```

Comment:
"one of `create_block*` methods" --> "one of the `create_block*` methods"
Besides, I am not good at coding but is * necessary in `create_block*`?

Reply:
`*` denotes that there are several methods in `TestKit` starting with `create_block`, rather than a single method.